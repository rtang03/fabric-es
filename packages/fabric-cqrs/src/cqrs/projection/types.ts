import { Commit } from '../../types';

export interface FindAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      where?: Record<string, any>;
      all?: boolean;
      contain?: string | number;
    };
  };
}

export interface UpsertAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      commit: Record<string, Commit>;
    };
  };
}

export interface UpsertManyAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      commits: Record<string, Commit>;
    };
  };
}
