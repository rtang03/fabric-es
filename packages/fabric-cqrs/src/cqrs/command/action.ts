import { getAction, getErrorAction, getSuccessAction } from '../utils';
import {
  CreateAction,
  DeleteByEntityIdAction,
  DeleteByEntityIdCommitIdAction,
  QueryByEntIdCommitIdAction,
  QueryByEntityIdAction,
  QueryByEntityNameAction
} from './types';

const CREATE = '[Entity-C] Create';
const CREATE_ERROR = '[Entity-C] Create Error';
const CREATE_SUCCESS = '[Entity-C] Create Success';
const DELETE_BY_ENTITYID = '[Entity-C] Delete entities By entityName and id';
const DELETE_ERROR = '[Entity-C] Delete Error';
const DELETE_SUCCESS = '[Entity-C] Delete Success';
const DELETE_BY_ENTITYID_COMMITID =
  '[Entity-C] Delete By entityId and commitId';
const QUERY_BY_ENTITY_ID = '[Entity-C] Query entity by entity id';
const QUERY_BY_ENTITY_NAME = '[Entity-C] Query entity by entityName';
const QUERY_BY_ENTITYID_COMMITID = '[Entity-C] Query by entity id and commitId';
const QUERY_ERROR = '[Entity-C] Query Error';
const QUERY_SUCCESS = '[Entity-C] Query Success';

export const action = {
  CREATE,
  CREATE_ERROR,
  CREATE_SUCCESS,
  DELETE_BY_ENTITYID,
  DELETE_ERROR,
  DELETE_SUCCESS,
  DELETE_BY_ENTITYID_COMMITID,
  QUERY_BY_ENTITY_ID,
  QUERY_BY_ENTITY_NAME,
  QUERY_BY_ENTITYID_COMMITID,
  QUERY_ERROR,
  QUERY_SUCCESS,
  create: getAction<CreateAction>(CREATE),
  deleteByEntityId: getAction<DeleteByEntityIdAction>(DELETE_BY_ENTITYID),
  deleteByEntityIdCommitId: getAction<DeleteByEntityIdCommitIdAction>(
    DELETE_BY_ENTITYID_COMMITID
  ),
  queryByEntityId: getAction<QueryByEntityIdAction>(QUERY_BY_ENTITY_ID),
  queryByEntityName: getAction<QueryByEntityNameAction>(QUERY_BY_ENTITY_NAME),
  queryByEntIdCommitId: getAction<QueryByEntIdCommitIdAction>(
    QUERY_BY_ENTITYID_COMMITID
  ),
  createError: getErrorAction(CREATE_ERROR),
  createSuccess: getSuccessAction(CREATE_SUCCESS),
  deleteError: getErrorAction(DELETE_ERROR),
  deleteSuccess: getSuccessAction(DELETE_SUCCESS),
  queryError: getErrorAction(QUERY_ERROR),
  querySuccess: getSuccessAction(QUERY_SUCCESS)
};
