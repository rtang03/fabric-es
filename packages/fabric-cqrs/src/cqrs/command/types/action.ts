import { BaseEvent } from '../../../types';

export interface CreateAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      entityName: string;
      id: string;
      version: number;
      events: BaseEvent[];
      collection?: string;
    };
    enrollmentId?: string;
  };
}

export interface DeleteByEntityIdAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      entityName: string;
      id: string;
      collection?: string;
    };
  };
}

export interface DeleteByEntityIdCommitIdAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      entityName: string;
      id: string;
      commitId: string;
      collection?: string;
    };
  };
}

export interface QueryByEntityIdAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      entityName: string;
      id: string;
      collection?: string;
    };
  };
}

export interface QueryByEntityNameAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      entityName: string;
      collection?: string;
    };
  };
}

export interface QueryByEntIdCommitIdAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      entityName: string;
      id: string;
      commitId: string;
      collection?: string;
    };
  };
}

export type CommandActions =
  | CreateAction
  | DeleteByEntityIdAction
  | DeleteByEntityIdCommitIdAction
  | QueryByEntityIdAction
  | QueryByEntityNameAction
  | QueryByEntIdCommitIdAction;
