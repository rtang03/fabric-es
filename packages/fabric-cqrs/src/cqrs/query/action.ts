/**
 * @packageDocumentation
 * @hidden
 */
import { getAction, getErrorAction, getSuccessAction } from '../utils';
import {
  DeleteByEntityIdAction,
  DeleteByEntityNameAction,
  MergeAction,
  MergeBatchAction,
  QueryByEntityIdAction,
  QueryByEntityNameAction
} from './types';

const DELETE_BY_ENTITYNAME = '[Entity-Q] Delete entities by EntityName';
const DELETE_BY_ENTITY_ID = '[Entity-Q] Delete entities by entityId';
const DELETE_ERROR = '[Entity-Q] Delete Error';
const DELETE_SUCCESS = '[Entity-Q] Delete Success';
const QUERY_BY_ENTITYNAME = '[Entity-Q] Query entities By EntityName';
const QUERY_BY_ENTITY_ID = '[Entity-Q] Query entities By entityId';
const QUERY_ERROR = '[Entity-Q] Query Error';
const QUERY_SUCCESS = '[Entity-Q] Query Success';
const MERGE = '[Entity-Q] Merge entity';
const MERGE_ERROR = '[Entity-Q] Merge Error';
const MERGE_SUCCESS = '[Entity-Q] Merge Success';
const MERGE_BATCH = '[Entity-Q] Merge records of entities';
const MERGE_BATCH_ERROR = '[Entity-Q] Merge Batch Error';
const MERGE_BATCH_SUCCESS = '[Entity-Q] Merge Batch Success';

export const action = {
  DELETE_BY_ENTITYNAME,
  DELETE_BY_ENTITY_ID,
  DELETE_ERROR,
  DELETE_SUCCESS,
  QUERY_BY_ENTITYNAME,
  QUERY_BY_ENTITY_ID,
  QUERY_ERROR,
  QUERY_SUCCESS,
  MERGE,
  MERGE_ERROR,
  MERGE_SUCCESS,
  MERGE_BATCH,
  MERGE_BATCH_ERROR,
  MERGE_BATCH_SUCCESS,
  deleteByEntityId: getAction<DeleteByEntityIdAction>(DELETE_BY_ENTITY_ID),
  deleteByEntityName: getAction<DeleteByEntityNameAction>(DELETE_BY_ENTITYNAME),
  merge: getAction<MergeAction>(MERGE),
  mergeBatch: getAction<MergeBatchAction>(MERGE_BATCH),
  queryByEntityName: getAction<QueryByEntityNameAction>(QUERY_BY_ENTITYNAME),
  queryByEntityId: getAction<QueryByEntityIdAction>(QUERY_BY_ENTITY_ID),
  deleteError: getErrorAction(DELETE_ERROR),
  deleteSuccess: getSuccessAction(DELETE_SUCCESS),
  queryError: getErrorAction(QUERY_ERROR),
  querySuccess: getSuccessAction(QUERY_SUCCESS),
  mergeError: getErrorAction(MERGE_ERROR),
  mergeSuccess: getSuccessAction(MERGE_SUCCESS),
  mergeBatchError: getErrorAction(MERGE_BATCH_ERROR),
  mergeBatchSuccess: getSuccessAction(MERGE_BATCH_SUCCESS)
};
