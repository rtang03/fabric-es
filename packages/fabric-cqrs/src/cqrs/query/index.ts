/**
 * @packageDocumentation
 * @hidden
 */
import { epic as queryEpic } from './epic';

export { action } from './action';
export { reducer } from './reducer';
export { queryEpic };
