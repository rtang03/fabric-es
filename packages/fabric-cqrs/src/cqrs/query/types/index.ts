/**
 * @packageDocumentation
 * @hidden
 */
export * from './action';
