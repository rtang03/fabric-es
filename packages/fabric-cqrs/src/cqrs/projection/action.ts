/**
 * @packageDocumentation
 * @hidden
 */
import { getAction, getSuccessAction } from '../utils';
import { FindAction, UpsertAction, UpsertManyAction } from './types';

const UPSERT = '[Project] Upsert Entity';
const UPSERT_SUCCESS = '[Project] Upsert Entity Success';
const UPSERT_MANY = '[Project] Upsert Many Entity';
const UPSERT_MANY_SUCCESS = '[Project] Upsert Many Success';
const FIND = '[Project] Find';
const FIND_SUCCESS = '[Project] Find Success';

export const action = {
  UPSERT: MERGE,
  UPSERT_SUCCESS: MERGE_ENTITY_SUCCESS,
  UPSERT_MANY: MERGE_ENTITY_BATCH,
  UPSERT_MANY_SUCCESS: MERGE_ENTITY_BATCH_SUCCESS,
  FIND: MERGE_ENTITY_ERROR,
  FIND_SUCCESS: MERGE_ENTITY_BATCH_ERROR,
  find: getAction<FindAction>(FIND),
  upsert: getAction<UpsertAction>(UPSERT),
  mergeEntityBatch: getAction<UpsertManyAction>(UPSERT_MANY),
  mergeEntityBatchError: getSuccessAction(FIND_SUCCESS),
  mergeEntitySuccess: getSuccessAction(UPSERT_SUCCESS),
  mergeEntityBatchSuccess: getSuccessAction(UPSERT_MANY_SUCCESS)
};
