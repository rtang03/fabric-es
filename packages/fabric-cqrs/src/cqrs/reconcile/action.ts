/**
 * @packageDocumentation
 * @hidden
 */
import { getAction, getErrorAction, getSuccessAction } from '../utils';
import { MergeAction, ReconcileAction } from './types';

const RECONCILE = '[Entity] Reconcile';
const RECONCILE_ERROR = '[Entity] Reconcile Error';
const RECONCILE_SUCCESS = '[Entity] Reconcile Success';
const MERGE = '[Entity] Merge the queried entity to Query-side';

export const action = {
  RECONCILE,
  RECONCILE_ERROR,
  RECONCILE_SUCCESS,
  MERGE,
  reconcile: getAction<ReconcileAction>(RECONCILE),
  merge: getAction<MergeAction>(MERGE),
  reconcileError: getErrorAction(RECONCILE_ERROR),
  reconcileSuccess: getSuccessAction(RECONCILE_SUCCESS)
};
