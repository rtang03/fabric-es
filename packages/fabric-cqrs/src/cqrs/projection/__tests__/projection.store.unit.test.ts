import { Store } from 'redux';
import { reducer } from '../../../example';
import { createProjectionDb, createQueryDatabase } from '../../../peer';
import { Commit } from '../../../types';
import { action as queryAction } from '../../query';
import { generateToken } from '../../../store/utils';
import { action } from '../action';
import { getStore } from './__utils__/store';

let store: Store;
const context: any = {
  projectionDb: createProjectionDb('test-entity'),
  queryDatabase: createQueryDatabase(),
  reducer
};

beforeAll(() => {
  store = getStore(context);
});

describe('CQRS - projection Tests', () => {
  it('should upsert many', done => {
    const commits: Record<string, Commit> = {
      '001': {
        id: '1000',
        entityName: 'test-entity',
        events: [{ type: 'ADD' }]
      },
      '002': {
        id: '1000',
        entityName: 'test-entity',
        events: [{ type: 'ADD' }]
      },
      '003': {
        id: '1000',
        entityName: 'test-entity',
        events: [{ type: 'MINUS' }]
      },
      '004': {
        id: '2000',
        entityName: 'test-entity',
        events: [{ type: 'ADD' }]
      },
      '005': {
        id: '2000',
        entityName: 'test-entity',
        events: [{ type: 'ADD' }]
      }
    };
    const tid = generateToken();
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.UPSERT_MANY_SUCCESS) {
        expect(result).toMatchSnapshot();
        unsubscribe();
        done();
      }
    });
    store.dispatch(action.upsertMany({ tx_id: tid, args: { commits }, store }));
  });

  it('should find #1 by All', done => {
    const tid = generateToken();
    const all = true;
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.FIND_SUCCESS) {
        expect(result).toMatchSnapshot();
        unsubscribe();
        done();
      }
    });
    store.dispatch(action.find({ tx_id: tid, args: { all }, store }));
  });

  it('should find #2 by where', done => {
    const tid = generateToken();
    const where = { value: 2 };
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.FIND_SUCCESS) {
        expect(result).toMatchSnapshot();
        unsubscribe();
        done();
      }
    });
    store.dispatch(action.find({ tx_id: tid, args: { where }, store }));
  });

  it('should upsert', done => {
    const tid = generateToken();
    const commit: Record<string, Commit> = {
      '1001': {
        id: '91000',
        entityName: 'test-entity',
        events: [{ type: 'ADD' }]
      },
      '1002': {
        id: '91000',
        entityName: 'test-entity',
        events: [{ type: 'ADD' }]
      },
      '1003': {
        id: '91000',
        entityName: 'test-entity',
        events: [{ type: 'MINUS' }]
      }
    };
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.UPSERT_SUCCESS) {
        expect(result).toMatchSnapshot();
        unsubscribe();
        done();
      }
    });
    store.dispatch(action.upsert({ tx_id: tid, args: { commit }, store }));
  });

  it('should find #3 by contain', done => {
    const tid = generateToken();
    const contain = '9';
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.FIND_SUCCESS) {
        expect(result).toMatchSnapshot();
        unsubscribe();
        done();
      }
    });
    store.dispatch(action.find({ tx_id: tid, args: { contain }, store }));
  });

  it('should do projection, when QueryDB is merged', done => {
    const tid = generateToken();
    const commit: Commit = {
      id: '101000',
      commitId: '20181208155814607',
      entityName: 'test-entity',
      entityId: 'ent_test_1001',
      version: 0,
      events: [{ type: 'ADD' }]
    };
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.UPSERT_SUCCESS) {
        expect(result).toEqual({ '101000': {} });
        unsubscribe();
        done();
      }
    });
    store.dispatch(queryAction.merge({ tx_id: tid, args: { commit }, store }));
  });

  it('should find #4 by where', done => {
    const tid = generateToken();
    const where = { id: '101000' };
    const unsubscribe = store.subscribe(() => {
      const { tx_id, result, type } = store.getState().projection;
      if (tx_id === tid && type === action.FIND_SUCCESS) {
        expect(result).toEqual([{ id: '101000', value: 1 }]);
        unsubscribe();
        done();
      }
    });
    store.dispatch(action.find({ tx_id: tid, args: { where }, store }));
  });
});
