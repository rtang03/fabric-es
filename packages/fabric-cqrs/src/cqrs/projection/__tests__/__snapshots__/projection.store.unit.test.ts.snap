// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CQRS - projection Tests should find #1 by All 1`] = `
Array [
  Object {
    "id": "1000",
    "value": 1,
  },
  Object {
    "id": "2000",
    "value": 2,
  },
]
`;

exports[`CQRS - projection Tests should find #2 by where 1`] = `
Array [
  Object {
    "id": "2000",
    "value": 2,
  },
]
`;

exports[`CQRS - projection Tests should find #3 by contain 1`] = `
Array [
  Object {
    "id": "91000",
    "value": 1,
  },
]
`;

exports[`CQRS - projection Tests should upsert 1`] = `
Object {
  "91000": Object {},
}
`;

exports[`CQRS - projection Tests should upsert many 1`] = `
Object {
  "1000": Object {},
  "2000": Object {},
}
`;
